pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -508,6 +508,18 @@ func (s *DiscoveryServer) edsUpdate(shard, serviceName string,
 	defer s.mutex.Unlock()
 	requireFull := false
 
+	// To prevent memory leak.
+	// Should delete the service EndpointShards, when endpoints deleted or service deleted.
+	if len(istioEndpoints) == 0 {
+		if s.EndpointShardsByService[serviceName] != nil {
+			delete(s.EndpointShardsByService[serviceName].Shards, shard)
```

Comment:
`s.EndpointShardsByService[serviceName].Lock()` is needed before deleting the shard.

Reply:
good catch